files/routes/votes.py

```diff
@@ -56,62 +56,72 @@ def admin_vote_info_get(v):
 @is_not_permabanned
 def api_vote_post(post_id, new, v):
 
-	if new == "-1" and environ.get('DISABLE_DOWNVOTES') == '1': return {"error": "forbidden."}, 403
-
-	if new not in ["-1", "0", "1"]: abort(400)
-
+	# make sure we're allowed in (is this really necessary? I'm not sure)
 	if request.headers.get("Authorization"): abort(403)
 
+	# make sure new is valid
+	if new == "-1" and environ.get('DISABLE_DOWNVOTES') == '1': return {"error": "forbidden."}, 403
+	if new not in ["-1", "0", "1"]: abort(400)
 	new = int(new)
 
+	# get the post
+	try: post_id = int(post_id)
+	except: abort(404)
 	post = get_post(post_id)
 
-	existing = g.db.query(Vote).filter_by(user_id=v.id, submission_id=post.id).one_or_none()
+	# verify that the post is allowed to be voted on
+	if post.author_id in {AUTOPOLLER_ID,AUTOBETTER_ID,AUTOCHOICE_ID}: return {"error": "forbidden."}, 403
+	
+	# get the old vote, if we have one
+	vote = g.db.query(Vote).filter_by(user_id=v.id, submission_id=post.id).one_or_none()
 
-	coin_delta = 1
-	if v.id == post.author.id:
-		coin_delta = 0
+	# should we just do nothing? we could just do nothing
+	if vote and vote.vote_type == new: return "", 204
 
-	if existing and existing.vote_type == new: return "", 204
+	# at this point we are guaranteed to be making a change
 
-	if existing:
-		if existing.vote_type == 0 and new != 0:
-			post.author.coins += coin_delta
-			post.author.truecoins += coin_delta
-			g.db.add(post.author)
-			existing.vote_type = new
-			g.db.add(existing)
-		elif existing.vote_type != 0 and new == 0:
-			post.author.coins -= coin_delta
-			post.author.truecoins -= coin_delta
-			g.db.add(post.author)
-			g.db.delete(existing)
-		else:
-			existing.vote_type = new
-			g.db.add(existing)
-	elif new != 0:
-		post.author.coins += coin_delta
-		post.author.truecoins += coin_delta
-		g.db.add(post.author)
+	# author votes don't matter if it's the author themselves
+	points_matter = (v.id != post.author.id)
 
+	if vote:
+		# remove the old score data
+		if points_matter:
+			post.author.coins -= vote.vote_type
+			post.author.truecoins -= vote.vote_type
+			# we'll be saving later anyway, so don't bother doing so here
+	else:
+		# create new vote data
 		DEFAULT_IMAGE = '/assets/images/default-profile-pic.webp'
```

Comment:
Not really sure what they were doing with this but I guess we don't need it anymore

Reply:
I think it was just a really ugly logic ball intended to only change the values once.